Implement secure browser storage

melissa-barca opened the pull request:
Addresses https://github.com/rstudio/vscode-server/issues/174

This will need to make it's way in to [rstudio/vscode-server](https://github.com/rstudio/vscode-server), but for ease-of-development I've implemented this in the Positron repo.

Desktop has a secret storage service implemented via electron that relies on the OS-specific encryption service. Before this PR, there was no server encryption service and encrypted data was only stored local to the active session and did not survive a page reload.
This PR implements a Browser Encryption Service using the Web Crypto API. 

- The service encrypts user data using AES-GCM encryption via a user-specific encryption key and saves the encrypted data in the browser storage
- The service derives a wrapping key using PBKDF2 from a randomly generated user salt value and stores the wrapped encryption key in browser storage for use between sessions.
- The salt is stored on the file system in the user's positron data directory as a hidden file named `.positron.salt`. If the filesystem cannot be written to, the service fails and doesn't store the data.
- Data saved in the browser storage will also be written to Workbench's internal database when run though Workbench

### Release Notes

<!--
  Optionally, replace `N/A` with text to be included in the next release notes.
  The `N/A` bullets are ignored. If you refer to one or more Positron issues,
  these issues are used to collect information about the feature or bugfix, such
  as the relevant language pack as determined by Github labels of type `lang: `.
  The note will automatically be tagged with the language.

  These notes are typically filled by the Positron team. If you are an external
  contributor, you may ignore this section.
-->

#### New Features

- N/A

#### Bug Fixes

- N/A


### QA Notes

@:web
@:quarto

melissa-barca committed:
first pass on browser encryption service

melissa-barca committed:
code cleanup

melissa-barca committed:
refactor for code cleanup

melissa-barca committed:
use salt from user's file system to wrap the encryption key before
saving it

jmcphers commented on src/vs/workbench/services/encryption/browser/browserEncryptionService.ts:
```diff
@@ -0,0 +1,315 @@
+		}
+
+		salt = mainWindow.crypto.getRandomValues(new Uint8Array(16));
+		this.storageService.store(this._storageSaltKey, JSON.stringify(Array.from(salt)), StorageScope.APPLICATION, StorageTarget.MACHINE);
```
It's good to have a fallback here, but as I'm reading it, this stores the salt and the wrapped key in the same place, which is the same place a lot of values will also be stored, so if someone gets access to the encrypted values they'll also get everything needed to decrypt them.

Maybe better to just fail if we can't use the salt file. What do you think?

melissa-barca replied:
Your assessment is correct. I wondered about this too and couldn't decide. It also can get us in a weird place if we temporarily cannot write to the file system but gain access in a future session (we'll generate a new key, but the old key and encrypted values will still exist in browser storage, but attempts to decrypt the values will fail). I'll remove this and just fail out.

timtmok commented on src/vs/workbench/services/encryption/browser/browserEncryptionService.ts:
```diff
@@ -0,0 +1,315 @@
+	}
+
+	setUsePlainTextEncryption(): Promise<void> {
+		this.logService.info('[BrowserEncryptionService] Setting use plain text encryption');
```
It seems to be a Linux only capability for Electron: https://www.electronjs.org/docs/latest/api/safe-storage#safestoragesetuseplaintextencryptionuseplaintext

Given that `encryptionService.ts` is the same, I think this is the intended implementation for the web?

melissa-barca replied:
Right, this is a fallback method for when OS encryption isn't available in desktop. I added some additional comments to clarify.

melissa-barca committed:
move salt into a subfolder, implement DeferredPromise pattern and
add context to missing functions, remove fallback salt browser
storage, and improve logging

melissa-barca committed:
Merge branch 'main' into secure-storage

Signed-off-by: Melissa Barca <5323711+melissa-barca@users.noreply.github.com>

jmcphers reviewed (approved):
LG. The docs for `crypto.subtle` suggest you need to be in an HTTPS context, but I was able to see this working with regular HTTP and a dev build of Positron. Confirmed that secrets are stored and retrieved across sessions and workspaces.

jeffvroom reviewed (approved):
LGTM!

melissa-barca commented:
Closing this as I merged the changes upstream: https://github.com/rstudio/vscode-server/pull/179 and will open a separate PR for pulling downstream.
